compiler/x/codegen/OMRTreeEvaluator.cpp

```diff
@@ -2429,6 +2429,80 @@ TR::Register *OMR::X86::TreeEvaluator::andORStringEvaluator(TR::Node *node, TR::
    return resultReg;
    }
 
+TR::Register *OMR::X86::TreeEvaluator::OverflowCHKEvaluator(TR::Node *node, TR::CodeGenerator *cg)
+   {
+   /*
+    *overflowCHK
+    *   =>child1(Operation)
+    *   =>child2(Operand1)
+    *   =>child3(Operand2)
+    */
+   bool firstChildEvaluatedAlready = node->getFirstChild()->getRegister()? true : false;
+   cg->evaluate(node->getFirstChild());
+   TR::Block *overflowCatchBlock = NULL;
+   TR::Instruction *prevInst = cg->comp()->getAppendInstruction();
+   // we need to do the operation again for the following cases:
+   // 1. the operation was somehow optimized to other forms, like LEA for example
+   // 2. the child was evaluated already under a different treetop
+   if (!(prevInst->getOpCode()).modifiesOverflowFlag() || firstChildEvaluatedAlready ) 
+      {
+      TR::Register *reg1; 
+      TR::Register *reg2 = cg->evaluate(node->getThirdChild());
+      TR_X86OpCodes op;
+      bool nodeIs64Bit; 
+      switch (node->getOverflowCHKInfo())
```

Comment:
The design here of only checking the opcode for the overflow check could lead to some surprising behaviour - an operation might slip through the evaluator for a while until we find that we need to recreate the calculation when the code generator will then fall over. Shouldn't we do this check up front to fail fast if someone gives us an overflowchk with an unknown opcode. We might be able to avoid recreating the operation for a while, but once we need to then we are going to have a problem and I would rather fail fast with asserts on I think to say that the problem situation could occur.